Deploy gke scripts (#6)
sayakpaul/deploy-hf-tf-vision-models: hf_vision_model_tfserving_gke/README.md

1. `./install_gcloud.sh`
    - This downloads and installs `gcloud` tool, `docker-credential-gcr` and `gke-gcloud-auth-plugin` components from `gcloud` CLI. 
  
    - Skip this step if you already have `gcloud` installed on your system. 
      - When you skip, and if you have not installed `docker-credential-gcr` and `gke-gcloud-auth-plugin` components yet, you should install them manually using `gcloud`.

2. `./gcloud_auth.sh`
    - This authenticates the access from `gcloud` to `GCP` project. Here are variables that you might need to customize inside the script.
      - `GCP_PROJECT_ID`: GCP project ID to autheticate

    - Skip this step if you have already authenticated beforehand.

3. `./build_tfserving.sh`
    - This builds and pushes a custom TF Serving Docker image with 🤗 `transformers` ViT model. It basically follows the [official document](https://www.tensorflow.org/tfx/serving/docker) to build an image.
    - Here are variables that you might need to customize inside the script
        - `GCP_PROJECT_ID`: GCP project ID to push custom built TF Serving Docker image

>         - `BASE_IMAGE_TAG`: Custom image will be built on top of this base TF Serving Docker image. This is set to `gcr.io/gcp-ml-172005/tfs-resnet-cpu-opt` by default which is CPU optimized TF Serving Docker image

Reviewer: Do we want to add a note on how this image was built?

        - `MODEL_RELEASE_REPO`: GitHub repository where to download the released model. This should follow the format of `username/reponame`. This is set to `sayakpaul/deploy-hf-tf-vision-models` by default
        - `MODEL_RELEASE_TAG`: Tag name of the GitHub Release where to download the released model. This is set to `1.0` by default
        - `MODEL_RELEASE_FILE`: Model filename to download from the GitHub Release's Asset under `MODEL_RELEASE_REPO` and `MODEL_RELEASE_TAG`. This is set to `saved_model.tar.gz` by default

>         - `MODEL_NAME`: Model name to be exposed to client as an endpoint from TFServing. For instance, the endpoint follows the format of `http://IP_ADDRESS:PORT_NUMBER/v1/models/MODEL_NAME:predict`. This is set to `hf-vit` by default

Reviewer: What is set to `hf-vit` be default?

        - `VERSION`: Version information for the custom built TF Serving Docker image. For instance, the tag of the custom built TF Serving Docker image follows the format of `tfserving-MODEL_NAME:VERSION`. This is set to `latest` by default, so the tag will be formed as `tfserving-hf-vit:latest` by default.

4. `./provision_gke_cluster.sh`

>     - This provisions a GKE clusters with `gcloud container clusters` CLI. Here are variables that you might need to customize inside the script. Most of the values in this scripts are set with the optimal values according to [this experiments](https://github.com/deep-diver/ml-deployment-k8s-tfserving)

Reviewer: Let's also mention that the experiments were conducted with a ResNet50 model which is different from ViT.

    - There is a only experiment at the moment, but you can simply add more configurations under `.kube/experiments` by inherting base ones.
    
    - Here are variables that you might need to customize inside the script.
      - `GCP_PROJECT_ID`: GCP project ID where the GKE cluster is
      - `BASE_IMAGE_TAG`, `MODEL_NAME`, `VERSION`, `NEW_IMAGE_NAME`: Same values as in `./build_tfserving.sh` script
      - `GKE_CLUSTER_NAME`, `GKE_CLUSTER_ZONE`: Same values as in `./provision_gke_cluster.sh` script
      - `TARGET_EXPERIMENT`: Name of the overlay configurations under `.kube/experiments`. This is set to `8vCPU+16GB+inter_op4` by default.
      - `GKE_DEPLOYMENT_NAME`: Name of the `Deployment`. This is set to `tfs-server` by default. This is going to be used to check rollout status of the deployment. 

After all of these steps goes successfully, you will see the output similar to below. There are two parts. 

> - The first part shows the endpoints of the deployment. Port number `8500` is for `HTTP/1.1` based RESTful API while Port number `8501` is for `HTTP/2` based gRPC API. 

Reviewer: First part of?

> - The first part shows the endpoints of the deployment. Port number `8500` is for `HTTP/1.1` based RESTful API while Port number `8501` is for `HTTP/2` based gRPC API. 

Reviewer: ```suggestion
- The first part shows the endpoints of the deployment. Port number `8500` is for `HTTP/1.1` based RESTful API, while port number `8501` is for `HTTP/2` based gRPC API. 
```

> - The second part shows the rolling status of the deployment. Pay attention that `Image` is set correctly, and there are two `Ports` for `HTTP/1.1` and `HTTP/2`. Also, some TF Serving specific flags are set in the `Args` (i.e `tensorflow_inter_op_parallelism` and `tensorflow_intra_op_parallelism`). Finally `Replicas` shows there are desired number of pods running. 

Reviewer: I think you're referring to `Name` by `Image` which is the GKE deployment name?

      - `MEM_CAPACITY`: Memory capacity for each node(VM) of the GKE cluster in `MiB`. THis is set to `16384` by default which is 16GB RAM.

5. `./install_kustomize.sh`
    - This downloads and installs `Kustomize` tool under `.kube` directory.

6. `./deploy_tfserving_gke.sh`: 
    - This apply `Deployment` and `Service` to the GKE cluster using `Kustomize`. Essentially, the deployment will be applied using `kustomize build experiments/$TARGET_EXPERIMENT`. This means `Kustomize` constructs yamls by overlaying the base ones in `.kube/base` with specific ones in `.kube/experiments/$TARGET_EXPERIMENT`

    - There is a only experiment at the moment, but you can simply add more configurations under `.kube/experiments` by inherting base ones.
    
    - Here are variables that you might need to customize inside the script.
      - `GCP_PROJECT_ID`: GCP project ID where the GKE cluster is
      - `BASE_IMAGE_TAG`, `MODEL_NAME`, `VERSION`, `NEW_IMAGE_NAME`: Same values as in `./build_tfserving.sh` script
      - `GKE_CLUSTER_NAME`, `GKE_CLUSTER_ZONE`: Same values as in `./provision_gke_cluster.sh` script

>       - `TARGET_EXPERIMENT`: Name of the overlay configurations under `.kube/experiments`. This is set to `8vCPU+16GB+inter_op4` by default.

Reviewer: ```suggestion
      - `TARGET_EXPERIMENT`: Name of the overlay configurations under `.kube/experiments`. This is set to `8vCPU+16GB+inter_op4` by default. One can change this to `8vCPU+64GB+inter_op4_w_batch` or add a folder under `.kube/experiments` and use that. 
```

> # Deploy 🤗 `transformers` ViT model with TFServing on GKE

Reviewer: ```suggestion
# Deploy 🤗 `transformers` ViT model with TF Serving on GKE
```

> There are five shell scripts to deploy 🤗 `transformers` ViT model with TFServing on GKE. Some of them may be skipped depending on your local dev environment, but run all of them sequentially if you have a zero-base dev environment.

Reviewer: ```suggestion
There are five shell scripts to deploy 🤗 `transformers` ViT model with TF Serving on GKE. Some of them may be skipped depending on your local dev environment, but if you're starting from scratch, then run all the scripts sequentially.
```

2. `bash ./gcloud_auth.sh`: Authenticate the access from `gcloud` to your `GCP`
    - Skip if you have already authenticated `gcloud`.
    - If you have not authenticated `gcloud`, you should download Service Account key (JSON) and provide the path in `SA_KEY_FILE` variable. Also speicy your GCP project ID in `GCP_PROJECT_ID` variable.

3. `bash ./build_tfserving.sh`: Builds and pushes a custom TFServing Docker image with 🤗 `transformers` ViT model. 
    - `GCP_PROJECT_ID`: GCP project ID where you want to push a custom built TFServing Docker image
    - `BASE_IMAGE_TAG`: Base TFServing Docker image. Your custom image will be built on top of this.
    - `MODEL_RELEASE_REPO`: GitHub repository (`username/reponame`)
    - `MODEL_RELEASE_TAG`: GitHub Release tag 
    - `MODEL_RELEASE_FILE`: Asset filename to download from GitHub Release under `MODEL_RELEASE_REPO` and `MODEL_RELEASE_TAG`
    - `MODEL_NAME`: Model name to be exposed to client as an endpoint from TFServing
    - `VERSION`: Version information for the custom built TFServing Docker image

4. `bash ./install_kustomize.sh`: installs `Kustomize` CLI tool under `.kube` directory.

> 5. `bash ./provision_gke.sh`: provision `Deployment` and `Service` to the GKE cluster.

Reviewer: We also need to include instructions for setting up the GKE cluster.
